src/Dev/Validation/RelationValidationService.php

```diff
@@ -0,0 +1,648 @@
+<?php
+
+namespace SilverStripe\Dev\Validation;
+
+use ReflectionException;
+use SilverStripe\Core\ClassInfo;
+use SilverStripe\Core\Config\Configurable;
+use SilverStripe\Core\Flushable;
+use SilverStripe\Core\Injector\Injectable;
+use SilverStripe\Core\Resettable;
+use SilverStripe\ORM\DataObject;
+use SilverStripe\ORM\DB;
+
+/**
+ * Basic validation of relationship setup, this tool makes sure your relationships are set up correctly in both directions
+ * The validation is configurable and inspection can be narrowed down by namespace, class and relation name
+ *
+ * This tool is opt-in and runs via flush and outputs notices
+ * For strict validation it is recommended to hook this up to your unit test suite
+ */
+class RelationValidationService implements Flushable, Resettable
+{
+    use Configurable;
+    use Injectable;
+
+    /**
+     * Enable / disable validation output during flush
+     * This is disabled by default (opt-in)
+     *
+     * @var bool
+     */
+    private static $flush_output_enabled = true;
+
+    /**
+     * Only inspect classes with the following namespaces/class prefixes
+     * Empty string is a special value which represents classes without namespaces
+     * Set the value to null to disable the rule (useful when overriding configuration)
+     *
+     * @var array
+     */
+    private static $allow_rules = [
+        'empty' => '',
+        'app' => 'App',
+    ];
+
+    /**
+     * Any classes with the following namespaces/class prefixes will not be inspected
+     * This config is intended to be used together with @see $allow_rules to narrow down the inspected classes
+     * Empty string is a special value which represents classes without namespaces
+     * Set the value to null to disable the rule (useful when overriding configuration)
+     *
+     * @var array
+     */
+    private static $deny_rules = [];
+
+    /**
+     * Relations listed here will not be inspected
+     * Format is <class>.<relation>
+     * for example: Page::class.'.LinkTracking'
+     *
+     * @var array
+     */
+    private static $deny_relations = [];
+
+    /**
+     * Ignore any configuration, useful for debugging specific classes
+     *
+     * @var bool
+     */
+    protected $ignoreConfig = false;
+
+    /**
+     * @var array
+     */
+    protected $errors = [];
+
+    public function flushErrors(): void
```

Comment:
```suggestion
    public function clearErrors(): void
```
Probably best to avoid the word 'flush' since this is already a 'flusable' class